src/Compilers/CSharp/Portable/BoundTree/LengthBasedStringSwitchData.cs

```diff
@@ -0,0 +1,313 @@
+            foreach (var inputCase in inputCases)
+            {
+                if (inputCase.value.IsNull)
+                {
+                    Debug.Assert(nullCaseLabel is null, "At most one null case per string dispatch");
+                    nullCaseLabel = inputCase.label;
+                }
+            }
+
+            var lengthCaseLabels = ArrayBuilder<(ConstantValue value, LabelSymbol label)>.GetInstance();
+            var charJumpTables = ArrayBuilder<CharJumpTable>.GetInstance();
+            var stringJumpTables = ArrayBuilder<StringJumpTable>.GetInstance();
+            foreach (var group in inputCases.Where(c => !c.value.IsNull).GroupBy(c => c.value.StringValue!.Length))
+            {
+                int stringLength = group.Key;
+                var labelForLength = CreateAndRegisterCharJumpTables(stringLength, group.Select(e => (e.value, e.label)).ToImmutableArray(), charJumpTables, stringJumpTables);
+                lengthCaseLabels.Add((ConstantValue.Create(stringLength), labelForLength));
+            }
+
+            var lengthJumpTable = new LengthJumpTable(nullCaseLabel, lengthCaseLabels.ToImmutableAndFree());
+            return new LengthBasedStringSwitchData(lengthJumpTable, charJumpTables.ToImmutableAndFree(), stringJumpTables.ToImmutableAndFree());
+        }
+
+        private static LabelSymbol CreateAndRegisterCharJumpTables(int stringLength, ImmutableArray<(ConstantValue value, LabelSymbol label)> casesWithGivenLength,
+            ArrayBuilder<CharJumpTable> charJumpTables, ArrayBuilder<StringJumpTable> stringJumpTables)
+        {
+            Debug.Assert(stringLength >= 0);
+            Debug.Assert(casesWithGivenLength.All(c => c.value.StringValue!.Length == stringLength));
+            Debug.Assert(casesWithGivenLength.Length > 0);
+
+            if (stringLength == 0)
+            {
+                // Only the empty string has zero Length, no need for further testing
+                return casesWithGivenLength.Single().label;
+            }
+
+            if (casesWithGivenLength.Length == 1)
+            {
+                // We only have one case for the given string length, we don't need to do a char test
+                // Instead we'll jump straight to the final string test
+                return CreateAndRegisterStringJumpTable(casesWithGivenLength, stringJumpTables);
+            }
+
+            var bestCharacterPosition = selectBestCharacterIndex(stringLength, casesWithGivenLength);
+            var charCaseLabels = ArrayBuilder<(ConstantValue value, LabelSymbol label)>.GetInstance();
+            foreach (var group in casesWithGivenLength.GroupBy(c => c.value.StringValue![bestCharacterPosition]))
+            {
+                char character = group.Key;
+                var label = CreateAndRegisterStringJumpTable(group.ToImmutableArray(), stringJumpTables);
+                charCaseLabels.Add((ConstantValue.Create(character), label));
```

Comment:
>(ConstantValue.Create(character) [](http://example.com/codeflow?start=35&length=32)

Would it simplify anything if the values in `charCaseLabels` were `int` rather than `ConstantValue`? It looks like the `ConstantValue` is only used when calling `EmitIntegerSwitchJumpTable` where we're already mapping the array elements from `(ConstantValue, Label)` to `KeyValuePair<ConstantValue, Label>`.